scripts/generate_error_code.hs

```diff
@@ -0,0 +1,250 @@
+#!/usr/bin/env stack
+-- stack --resolver lts-17.3 script --package string-interpolate --package universum --package text-manipulate
+
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE NamedFieldPuns #-}
+
+-- | Script for adding new error to ligo and also update the docs.
+module GenerateErrorCode where
+
+import Data.String.Interpolate (i)
+import Data.Text.Manipulate (toCamel)
+import Prelude ()
+import Universum
+
+data ErrorItem = ErrorItem
+  { eiLabel :: Text
+  , eiCode :: Integer
+  , eiDesc :: Text
+  }
+
+data ShortErrorItem = Text :? Text
+  deriving (Eq)
+
+removedError :: ShortErrorItem
+removedError = "" :? ""
+
+errorsEnumerate :: Integer -> [ShortErrorItem] -> [ErrorItem]
+errorsEnumerate start =
+  map (\(eiCode, eiLabel :? eiDesc) -> ErrorItem{..}) .
+  filter ((/= removedError) . snd) .
+  zip [start..]
+
+
+invalidInputErrors :: [ErrorItem]
+invalidInputErrors = errorsEnumerate 100
+  -- New errors must be added to the _end_ of this list.
+  -- To remove an error, replace it with 'removedError'.
```

Comment:
May be I am missing something, but why can't we just use a `Map` from error code => (error_name, error_description) ?

AFAIU that will save us from putting these restrictions, that quite possibly could be missed when making an update?

Reply:
I took this from SegCFMM contract: https://github.com/serokell/segmented-cfmm/pull/25

There were some issues about conflicting error codes (2 PRs containing news error code), this would help solving that problem. Also `removedError` helps when we need to delete an error without changing error code of other errors.

Reply:
> There were some issues about conflicting error codes 

Ah, yes. That makes sense. So in that case, does that project include manually added tests to ensure that certain error codes map to the expected errors?

Reply:
Not sure why we want to test that. The script generate a haskell and ligo file to be used (example: https://github.com/tezos-commons/baseDAO/blob/rinn7e/%23235-numeric-error/haskell/src/Ligo/BaseDAO/ErrorCodes.hs)

We would want to use the constant from that file (Example: `notAdmin = 100`), instead of using `100` manually.

Reply:
> Not sure why we want to test that.

But what is preventing someone from not following the procedure and actually removing an error from the list (instead of replacing it with 'removed'?)

Reply:
Nothing really prevent them, but we can see clearly if we review the code (the error code changes for the errors below the removed error). Maybe @Martoon-00 could help explaining better on this.

Reply:
> but we can see clearly if we review the code..

Well, if this mixup happens (re-ordering of lines) happen in the final rebase, after the reviews, just before the merge, will it get caught?

Reply:
I'm not sure how this would happen but if 'removed` is re-order to a different line, we definitely can see the error codes change. I guess this require some attention to check before merging if the reviews are already done.

Reply:
> we definitely can see the error codes change

AFAIU you need to run a diff after running the generator to see it. But one can always just run the tests after running the generator, and proceed to commit and merge if everything passed in the tests, right? AFAIU it won't be caught in the CI as well.

Reply:
@sras Sorry, I missed your comment. What is your suggestion to fix this? Should we add a test to ensure the errors are map to proper error codes?

Reply:
@rinn7e Yes, If we explicitly write the expected error code for each error in the test, I think that should be enough.

Reply:
@sras I've added the tests, this should be good to go now.